src/Sarif.WorkItems/SarifWorkItemFiler.cs

```diff
@@ -0,0 +1,206 @@
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SarifWorkItemFiler"> class.</see>
+        /// </summary>
+        /// <param name="filingClient">
+        /// A client for communicating with a work item filing host (for example, GitHub or Azure DevOps).
+        /// </param>
+        /// <param name="filingContext">
+        /// A root context object that configures the work item filing operation.
+        /// </param>
+        public SarifWorkItemFiler(SarifWorkItemContext filingContext)
+        {
+            this.FilingContext = filingContext ?? throw new ArgumentNullException(nameof(filingContext));
+            this.FilingClient = FilingClientFactory.CreateFilingTarget(filingContext.ProjectUri);
+        }
+
+        public SarifWorkItemContext FilingContext { get; set; }
+
+        public FilingClient FilingClient { get; set; }
+
+        public virtual void FileWorkItems(Uri sarifLogFileLocation)
+        {
+            sarifLogFileLocation = sarifLogFileLocation ?? throw new ArgumentNullException(nameof(sarifLogFileLocation));
+
+            if (sarifLogFileLocation.IsAbsoluteUri && sarifLogFileLocation.Scheme == "file:")
+            {
+                FileWorkItems(File.ReadAllText(sarifLogFileLocation.LocalPath));
+            }
+
+            throw new NotImplementedException();
+        }
+
+        public virtual void FileWorkItems(string sarifLogFileContents)
+        {
+            sarifLogFileContents = sarifLogFileContents ?? throw new ArgumentNullException(nameof(sarifLogFileContents));
+
+            SarifLog sarifLog = JsonConvert.DeserializeObject<SarifLog>(sarifLogFileContents);
+            FileWorkItems(sarifLog);
+        }
+
+        public virtual void FileWorkItems(SarifLog sarifLog)
+        {
+            sarifLog = sarifLog ?? throw new ArgumentNullException(nameof(sarifLog));
+
+            this.FilingClient.Connect(this.FilingContext.SecurityToken).Wait();
+
+            OptionallyEmittedData optionallyEmittedData = this.FilingContext.DataToRemove;
+            if (optionallyEmittedData != OptionallyEmittedData.None)
+            {
+                var dataRemovingVisitor = new RemoveOptionalDataVisitor(optionallyEmittedData);
+                dataRemovingVisitor.Visit(sarifLog);
+            }
+
+            optionallyEmittedData = this.FilingContext.DataToInsert;
+            if (optionallyEmittedData != OptionallyEmittedData.None)
+            {
+                var dataInsertingVisitor = new InsertOptionalDataVisitor(optionallyEmittedData);
+                dataInsertingVisitor.Visit(sarifLog);
+            }
+
+            SplittingStrategy splittingStrategy = this.FilingContext.SplittingStrategy;
+            if (splittingStrategy == SplittingStrategy.None)
+            {
+                FileWorkItemsForSarifLog(sarifLog);
+                return;
+            }
+
+            for (int runIndex = 0; runIndex < sarifLog.Runs.Count; ++runIndex)
```

Comment:
>Runs [](start = 55, length = 4)

The spec allows `Runs` to be `null`, so check for this even before checking for `SplittingStrategy.None`.